### PR TITLE
[RL] Support Multi-Stage Awake

### LINKED ISSUE #3: Support for multi-instance memory saver
Currently, there's an assertion at https://github.com/fzyzcjy/torch_memory_saver/blob/master/torch_memory_saver/__init__.py#L18 to only allow one instance of memory saver. 

I was wondering what are the challenges involved in supporting multi-instance? 

Multi instance would provide several benefits. For eg, using a separate instance for KV cache and model weights in SGLang would allow more memory for complex RL workflows where you could resume the model weights memory region, update the weights from a copy, discard the copy and then resume the KV cache memory region.

### PR DESCRIPTION
**Pull request by hebiao064**
To Reviewer: Need this PR to check in first: https://github.com/fzyzcjy/torch_memory_saver/pull/17

Currently, many unit tests are failed because tms only support singleton without above change.

Related PR:

1. Torch Memory Saver: https://github.com/fzyzcjy/torch_memory_saver/pull/17
2. SGLang: https://github.com/sgl-project/sglang/pull/6962/files
3. https://github.com/volcengine/verl/pull/1911

Thanks a lot for @fzyzcjy 's guidance and help.

## Motivation
To close: https://github.com/fzyzcjy/torch_memory_saver/issues/3

vLLM has supported Multi-Stage Awake for vllm engine: https://github.com/vllm-project/vllm/issues/15254

But in SGLang, we are using `torch_memory_saver` for holding the model weight and KV Cache virtual address (to make sure CUDA Graph works across different rollouts)

And `torch_memory_saver` is a singleton, which make it hard for SGLang to support Multi-Stage Awake which is critical in RL use case

Without this feature, we can only set KV Cache mem frac rate to <= 0.7 or even lower (e.g 0.3)

Need this PR to check in first: https://github.com/fzyzcjy/torch_memory_saver/pull/17

## Checklist

- [x] Format your code according to the [Code Formatting with Pre-Commit](https://docs.sglang.ai/references/contribution_guide.html#code-formatting-with-pre-commit).
- [ ] Add unit tests as outlined in the [Running Unit Tests](https://docs.sglang.ai/references/contribution_guide.html#running-unit-tests-adding-to-ci).
- [ ] Update documentation / docstrings / example tutorials as needed, according to [Writing Documentation](https://docs.sglang.ai/references/contribution_guide.html#writing-documentation-running-docs-ci).
- [ ] Provide throughput / latency benchmark results and accuracy evaluation results as needed, according to [Benchmark and Profiling](https://docs.sglang.ai/references/benchmark_and_profiling.html) and [Accuracy Results](https://docs.sglang.ai/references/accuracy_evaluation.html).
- [ ] For reviewers: If you haven't made any contributions to this PR and are only assisting with merging the main branch, please remove yourself as a co-author when merging the PR.
- [ ] Please feel free to join our Slack channel at https://slack.sglang.ai to discuss your PR.

### CODE REVIEW

File: python/sglang/srt/managers/scheduler.py [hunk 1]
```diff
@@ -422,8 +422,13 @@ def __init__(
...
         # Init memory saver
-        self.memory_saver_adapter = TorchMemorySaverAdapter.create(
-            enable=server_args.enable_memory_saver
+        self.primary_memory_saver_adapter = TorchMemorySaverAdapter.create(
+            enable=server_args.enable_memory_saver,
+            is_primary=True,  # for model weight, we use primary memory saver
+        )
+        self.secondary_memory_saver_adapter = TorchMemorySaverAdapter.create(
+            enable=server_args.enable_memory_saver,
+            is_primary=False,  # for kv cache, we use secondary memory saver
```
Comment (zhaochenyang20) [inline python/sglang/srt/managers/scheduler.py]: should we use:

```python
TorchMemorySaverAdapter.create(
            enable=server_args.enable_memory_saver,
            release_target=["weights", "kv_cache"],
)
```

I do not know whether `release_target` is a good parameter name, but use a list could be clearer?

Comment (zhaochenyang20) [reply python/sglang/srt/managers/scheduler.py]: Better than "first" and "secondary"? hmmm

File: test/srt/test_release_memory_occupation.py [hunk 2]
```diff
@@ -7,27 +7,38 @@
...
+        """Common setup for engine and HF model."""
         model_name = DEFAULT_SMALL_MODEL_NAME_FOR_TEST
-        expect_output = " to spend it outdoors. I decided to"
-
         engine = sgl.Engine(
             model_path=model_name,
             random_seed=42,
             enable_memory_saver=True,
-            # disable_cuda_graph=True,  # for debugging only
+            disable_cuda_graph=True,  # for debugging only
```
Comment (zhaochenyang20) [inline test/srt/test_release_memory_occupation.py]: Since we want to test wether mem savor is compatible with cuda graph, why we disable it here?

Comment (zhaochenyang20) [reply test/srt/test_release_memory_occupation.py]: Also, could we explicitly add `mem_static = 0.9` here, to emphasize that two stage wake up support higher mem static?

**Review by zhaochenyang20 — CHANGES_REQUESTED** [review]

File: python/sglang/srt/managers/scheduler.py (hunk 1 above)
Comment (hebiao064) [reply python/sglang/srt/managers/scheduler.py]: we need different instance of Torch Memory Saver so passing a list might not be a good idea.

I can modify it to be 

```
TorchMemorySaverAdapter.create(
      enable=server_args.enable_memory_saver,
      id="kv_cache_memory_saver"
)

TorchMemorySaverAdapter.create(
      enable=server_args.enable_memory_saver,
      id="model_weights_memory_saver"
)
```

Let me know if it looks better to you.

File: test/srt/test_release_memory_occupation.py (hunk 2 above)
Comment (hebiao064) [reply test/srt/test_release_memory_occupation.py]: i forgot to uncomment it, will bring it back and set mem_frac_static to 0.9

Comment (hebiao064) [reply test/srt/test_release_memory_occupation.py]: done

File: python/sglang/srt/torch_memory_saver_adapter.py [hunk 3]
```diff
@@ -5,7 +5,17 @@
 try:
     import torch_memory_saver
 
+    # Primary Memory Saver for Model Weight
     _primary_memory_saver = torch_memory_saver.TorchMemorySaver()
+
+    # Secondary Memory Saver for KV Cache
+    _secondary_memory_saver = torch_memory_saver.TorchMemorySaver()
```
Comment (fzyzcjy) [inline python/sglang/srt/torch_memory_saver_adapter.py]: nit: wondering whether it would be good to make it more general, e.g. a dict of mem savers

File: python/sglang/srt/torch_memory_saver_adapter.py [hunk 4]
```diff
@@ -14,19 +24,52 @@
...
+
+        if enable:
+            if is_primary:
+                if _primary_adapter is None:
+                    _primary_adapter = _TorchMemorySaverAdapterReal(is_primary=True)
+                return _primary_adapter
+            else:
+                if _secondary_adapter is None:
+                    _secondary_adapter = _TorchMemorySaverAdapterReal(is_primary=False)
+                return _secondary_adapter
```
Comment (fzyzcjy) [inline python/sglang/srt/torch_memory_saver_adapter.py]: ... then we may be able to simplify code like this

**Review by fzyzcjy — COMMENTED** [review]
not yet fully reviewed, just spent a few minute to glance

File: python/sglang/srt/managers/scheduler.py (hunk 1 above)
Comment (hebiao064) [reply python/sglang/srt/managers/scheduler.py]: addressed feedback

File: python/sglang/srt/torch_memory_saver_adapter.py (hunk 3 above)
Comment (hebiao064) [reply python/sglang/srt/torch_memory_saver_adapter.py]: make sense, updated to dict

File: python/sglang/srt/torch_memory_saver_adapter.py (hunk 4 above)
Comment (hebiao064) [reply python/sglang/srt/torch_memory_saver_adapter.py]: make sense, updated to dict

File: python/sglang/srt/torch_memory_saver_adapter.py [hunk 5]
```diff
@@ -1,11 +1,32 @@
...
 from contextlib import contextmanager
 
 try:
     import torch_memory_saver
 
-    _primary_memory_saver = torch_memory_saver.TorchMemorySaver()
+    # Global lock to coordinate memory operations between instances
+    _memory_operation_lock = threading.Lock()
+
+    # Create memory savers with staggered timing to avoid address conflicts
```
Comment (fzyzcjy) [inline python/sglang/srt/torch_memory_saver_adapter.py]: hmm confused why we need this... (and getting worried is 0.01 enough in all cases)

File: python/sglang/srt/torch_memory_saver_adapter.py [hunk 6]
```diff
@@ -53,27 +123,63 @@ def enabled(self):
...
+        # Serialize resume operations and add retry with backoff
+        with _memory_operation_lock:
+            max_retries = 5
+            base_delay = 0.01
+
+            for attempt in range(max_retries):
+                try:
+                    # Add small delay between different instances
+                    if self.tag == "kv_cache":
+                        time.sleep(0.05)  # kv_cache waits a bit longer
```
Comment (fzyzcjy) [inline python/sglang/srt/torch_memory_saver_adapter.py]: (same)

File: python/sglang/srt/managers/scheduler.py [hunk 7]
```diff
@@ -2120,23 +2125,54 @@ def get_weights_by_name(self, recv_req: GetWeightsByNameReqInput):
...
+        tags = recv_req.tags
+        if tags is None:
+            # for backward compatibility, default to release both weights and kv cache
+            tags = ["weights", "kv_cache"]
+        if "weights" in tags:
+            self.stashed_model_static_state = _export_static_state(
+                self.tp_worker.worker.model_runner.model
+            )
+
+            self.weights_memory_saver_adapter.check_validity(
```
Comment (fzyzcjy) [inline python/sglang/srt/managers/scheduler.py]: nit: to be honest I feel the PR to be a little bit over complicated, if time permits, could you please update them a bit

**Review by fzyzcjy — COMMENTED** [review]
some nits

File: python/sglang/srt/torch_memory_saver_adapter.py (hunk 5 above)
Comment (hebiao064) [reply python/sglang/srt/torch_memory_saver_adapter.py]: removed, it was added to solving CUResult=1 issue, but didn't work

File: python/sglang/srt/managers/scheduler.py (hunk 7 above)
Comment (hebiao064) [reply python/sglang/srt/managers/scheduler.py]: simplified, let me know if its better now

File: python/sglang/srt/torch_memory_saver_adapter.py (hunk 6 above)
Comment (hebiao064) [reply python/sglang/srt/torch_memory_saver_adapter.py]: removed

File: python/sglang/srt/managers/scheduler.py [hunk 8]
```diff
@@ -422,8 +422,13 @@ def __init__(
...
         # Init memory saver
-        self.memory_saver_adapter = TorchMemorySaverAdapter.create(
-            enable=server_args.enable_memory_saver
+        self.weights_memory_saver_adapter = TorchMemorySaverAdapter.create(
+            enable=server_args.enable_memory_saver,
+            tag="weights",  # for model weight, we use weights memory saver
+        )
+        self.kv_cache_memory_saver_adapter = TorchMemorySaverAdapter.create(
+            enable=server_args.enable_memory_saver,
+            tag="kv_cache",  # for kv cache, we use kv_cache memory saver
```
Comment (zhaochenyang20) [inline python/sglang/srt/managers/scheduler.py]: nice naming

**Review by zhaochenyang20 — COMMENTED** [review]
LGTM

**Comment by hebiao064** [comment]
Will be using this PR: https://github.com/sgl-project/sglang/pull/7099

And here is the issue for tracking: https://github.com/sgl-project/sglang/issues/7009
